### PR TITLE
Pdfmake

### PR DESCRIPTION
**Pull request by ThomasJunk**
Basic implementation of infrastructure for pdfMake:

* SingleMotionContentProvider (motion/site.js)
 Encapsulates functionality to print a single motion
Has MotionContentProvider as a dependency.

* MotionContentProvider (motion/site.js)
Component to transform single motion to JSON 

* PdfMakeConverter (core/site.js)
Converts HTML to JSON objects to be consumed by pdfMake

* PdfMakeDocumentProvider (core/site.js)
Builds a document definition object to be consumed by pdfMake

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -22,6 +22,7 @@ Core:
 Motions:
 - Added origin field.
 - Added button to sort and number all motions in a category.
+- pdfMake for clientside generation of PDFs introduced
```
Comment (normanjaeckel) [inline CHANGELOG]: Proposal: `Introduced pdfMake for clientside generation of PDFs.`

File: openslides/core/migrations/0003_auto_20160815_1037.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
```
Comment (normanjaeckel) [inline openslides/core/migrations/0003_auto_20160815_1037.py]: Please rebase on top of master, then recreated all migration files.

File: openslides/motions/urls.py [hunk 3]
```diff
@@ -14,4 +14,5 @@
     url(r'^poll/(?P<poll_pk>\d+)/print/$',
         views.MotionPollPDF.as_view(),
         name='motionpoll_pdf'),
+    url(r'^encode_media/', views.encode_media, name="media_encoding")
```
Comment (normanjaeckel) [inline openslides/motions/urls.py]: Please add blank line befor this and a comma at the end of the line.

Mayba move this view to the mediafiles app.

File: openslides/motions/views.py [hunk 4]
```diff
@@ -466,3 +472,142 @@ def append_to_pdf(self, pdf):
             motions_to_pdf(pdf, motions)
         else:
             motion_to_pdf(pdf, self.get_object())
+
+
+@api_view(["POST"])
```
Comment (normanjaeckel) [inline openslides/motions/views.py]: Please use class based APIView form openslides.utils.view.

File: openslides/motions/views.py [hunk 5]
```diff
@@ -466,3 +472,142 @@ def append_to_pdf(self, pdf):
...
+def encode_media(request):
+    """
+    Encode_image is used in the context of PDF-Generation
+    Takes an array of IMG.src - Paths
+    Retrieves the according images
+    Encodes the images
```
Comment (normanjaeckel) [inline openslides/motions/views.py]: to BASE64?

File: openslides/motions/views.py [hunk 6]
```diff
@@ -466,3 +472,142 @@ def append_to_pdf(self, pdf):
...
+    return JsonResponse({
+        "images": images,
+        "fonts": fonts,
+        "defaultFont": default_font
+    })
+
+
+def get_default_font():
+    """
+    For development purposes this is hard coded
```
Comment (normanjaeckel) [inline openslides/motions/views.py]: Is this still required?

File: openslides/motions/views.py [hunk 7]
```diff
@@ -466,3 +472,142 @@ def append_to_pdf(self, pdf):
...
+    fonts = get_configured_fonts()
+
+    enc_fonts = [encode_font(name, files) for name, files in fonts.items()]
+
+    return enc_fonts
+
+
+def get_configured_fonts():
+    """
+    For development purposes, the current font definition is hard coded
```
Comment (normanjaeckel) [inline openslides/motions/views.py]: Here the same.

File: openslides/motions/views.py [hunk 8]
```diff
@@ -466,3 +472,142 @@ def append_to_pdf(self, pdf):
...
+    Returns the BASE64 encoded version of an image-file for a given path
+    :param file_path:
+    :return:
+    """
+    path = os.path.join(settings.MEDIA_ROOT, 'file', os.path.basename(file_path))
+    try:
+        with open(path, "rb") as file:
+            string_representation = "data:image/{};base64,{}".format(os.path.splitext(file_path)[1][1:],
+                                                                     base64.b64encode(file.read()).decode())
+    except:
```
Comment (normanjaeckel) [inline openslides/motions/views.py]: Please use explicit exception and add a comment what is happening in this case.

**Comment by emanuelschuetze** [comment]
Tested with Firefox:
* Submitter and status missing in PDF
* PDF should get a file name (displayed in tab title and used for save as)
* Import motion without <p>...</p>. Create pdf shows syntax error. Workaround: Open text in form. Save without changes. Editor added <p>..</p> around text. PDF works.

File: openslides/motions/static/templates/motions/motion-detail.html [hunk 9]
```diff
@@ -5,9 +5,9 @@
         <i class="fa fa-angle-double-left fa-lg"></i>
         <translate>All motions</translate>
       </a>
-      <a ui-sref="motions_single_pdf({pk: motion.id})" target="_blank" class="btn btn-default btn-sm">
+      <a ng-click="makePDF()" class="btn btn-primary btn-sm">
         <i class="fa fa-file-pdf-o fa-lg"></i>
-        <translate>PDF</translate>
+        <translate>PDFmake</translate>
```
Comment (emanuelschuetze) [inline openslides/motions/static/templates/motions/motion-detail.html]: Use "PDF" instead of "PDFmake".

**Comment by tsiegleauq** [comment]
Following diff fixes the problem with the broken header.
Also makes the definition of the header more comprehensible 

```
diff --git a/openslides/core/static/js/core/site.js b/openslides/core/static/js/core/site.js
index 522d1b8..7c8a9d9 100644
--- a/openslides/core/static/js/core/site.js
+++ b/openslides/core/static/js/core/site.js
@@ -36,10 +36,16 @@
                         return {
                             // alignment: 'center',
                             color: '#555',
-                            fontSize: 10,
                             margin: [80, 50, 80, 0], //margin: [left, top, right, bottom]
-                            columns: ['OpenSlides | Presentation and assembly system', {
+                            columns: [
+                            {
+                                fontSize: 10,
+                                width: '70%',
+                                text: 'OpenSlides | Presentation and assembly system',
+                            },
+                            {
                                 fontSize: 6,
+                                width: '30%',
                                 text: 'Stand: ' + date.toLocaleDateString() + " " + date.toLocaleTimeString(),
                                 alignment: 'right'
                             }]
```

**Comment by tsiegleauq** [comment]
Following diff sets the filename and uses download instead of open

```
diff --git a/openslides/motions/static/js/motions/site.js b/openslides/motions/static/js/motions/site.js
index 030349f..25ff478 100644
--- a/openslides/motions/static/js/motions/site.js
+++ b/openslides/motions/static/js/motions/site.js
@@ -749,11 +749,13 @@
         'MotionContentProvider',
         'PdfMakeConverter',
         'PdfMakeDocumentProvider',
+        'gettextCatalog',
         function($scope, $http, ngDialog,
             MotionForm, Motion, Category,
             Mediafile, Tag, User, Workflow,
             motion, SingleMotionContentProvider,
-            MotionContentProvider, PdfMakeConverter, PdfMakeDocumentProvider) {
+            MotionContentProvider, PdfMakeConverter, PdfMakeDocumentProvider,
+            gettextCatalog) {
             Motion.bindOne(motion.id, $scope, 'motion');
             Category.bindAll({}, $scope, 'categories');
             Mediafile.bindAll({}, $scope, 'mediafiles');
@@ -842,6 +844,7 @@
 
             $scope.makePDF = function() {
                 var content = motion.getText($scope.version) + motion.getReason($scope.version),
+                    id = motion.identifier,
                     slice = Function.prototype.call.bind([].slice),
                     map = Function.prototype.call.bind([].map),
                     image_sources = map($(content).find("img"), function(element) {
@@ -860,8 +863,9 @@
                     var converter = PdfMakeConverter.createInstance(data.images, data.fonts, pdfMake),
                         motionContentProvider = MotionContentProvider.createInstance(converter),
                         contentProvider = SingleMotionContentProvider.createInstance(motionContentProvider, motion, $scope, User),
-                        documentProvider = PdfMakeDocumentProvider.createInstance(contentProvider, data.defaultFont);
-                    pdfMake.createPdf(documentProvider.getDocument()).open();
+                        documentProvider = PdfMakeDocumentProvider.createInstance(contentProvider, data.defaultFont),
+                        filename = gettextCatalog.getString("Motion") + " " + id + ".pdf";
```

**Comment by tsiegleauq** [comment]
following diff re-adds the submitter name and the state

```
diff --git a/openslides/motions/static/js/motions/site.js b/openslides/motions/static/js/motions/site.js
index 030349f..e128f42 100644
--- a/openslides/motions/static/js/motions/site.js
+++ b/openslides/motions/static/js/motions/site.js
@@ -203,9 +203,11 @@
                      */
                     signment = function(motion, $scope, User) {
                         var label = converter.createElement("text", gettextCatalog.getString('Submitter') + ':\nStatus:');
+                        var state = converter.createElement("text", User.get(motion.submitters_id[0]).full_name + '\n'+gettextCatalog.getString(motion.state.name));
                         label.width = "30%";
                         label.bold = true;
-                        var signment = converter.createElement("stack", [label]);
+                        state.width = "70%";
+                        var signment = converter.createElement("columns", [label,state]);
                         signment.margin = [10, 20, 0, 10];
                         signment.lineHeight = 2.5;
                         return signment;
```

**Comment by tsiegleauq** [comment]
in core js, 
`ParseHtml = function(converted, htmlText) { ... }`

creates a "prevObject" when no tag was found during
`var html = $(htmlText.replace(/\t/g, "").replace(/\n/g, ""));`

(for example, when the HTML code is not surrounded by a <p> or <div>)

since this is not really an object, parsing for it using 
`slice(html).forEach(function(element)`

fails.

**Comment by emanuelschuetze** [comment]
Ok. Thanks!

**Comment by ThomasJunk** [comment]
Open Issue see #2287

**Pull request merged**